### ISSUE TITLE
predict() fails with ns() in conditional and dispersion formula.

### ISSUE THREAD

**Issue by tjmahr** [glmmTMB/glmmTMB#845]
Overlaps with #70 and #632.

Here's a minimal example. I use `ns()` in both the conditional formula and dispersion formula with different degrees of freedom. `predict()` doesn't work unless the two `ns()` have different function calls (e.g., ns() versus splines::ns()).

``` r
library(glmmTMB)
library(splines)

x <- glmmTMB(
  am ~ ns(wt, df = 3), 
  dispformula = ~ ns(wt, df = 2), 
  data = mtcars
)
newdata <- data.frame(
  wt = seq(min(mtcars$wt), max(mtcars$wt), length.out = 50)
)
predict(x, newdata = newdata)
#> Error in value[[jj]][ri, ] <- xij: number of items to replace is not a multiple of replacement length

# Don't use the same literal function in the dispformula
x <- glmmTMB(
  am ~ ns(wt, df = 3), 
  dispformula = ~ splines::ns(wt, df = 2), 
  data = mtcars
)
newdata <- data.frame(
  wt = seq(min(mtcars$wt), max(mtcars$wt), length.out = 50)
)
predict(x, newdata = newdata)
#>  [1]  1.001491e+00  9.990882e-01  9.964032e-01  9.931546e-01  9.890604e-01
#>  [6]  9.838390e-01  9.772084e-01  9.688868e-01  9.585924e-01  9.460434e-01
#> [11]  9.309579e-01  9.130541e-01  8.920501e-01  8.676643e-01  8.396146e-01
#> [16]  8.076193e-01  7.713966e-01  7.307445e-01  6.862062e-01  6.387312e-01
#> [21]  5.892730e-01  5.387849e-01  4.882205e-01  4.385332e-01  3.906765e-01
#> [26]  3.456036e-01  3.040200e-01  2.659067e-01  2.311128e-01  1.994874e-01
#> [31]  1.708796e-01  1.451386e-01  1.221134e-01  1.016533e-01  8.360722e-02
#> [36]  6.782437e-02  5.415387e-02  4.244482e-02  3.254636e-02  2.430758e-02
#> [41]  1.757763e-02  1.220560e-02  8.040627e-03  4.931824e-03  2.728309e-03
#> [46]  1.279202e-03  4.336202e-04  4.068282e-05 -5.049172e-05  9.215170e-06

```

My use case is a beta regression model where the mean changes with age and the dispersion around that mean changes with age and I want to allow some more flexibility to the mean (more knots) than the dispersion.

Fortunately, my workaround for predictions produces correct answers, compared to hardcoded spline functions.

``` r
# hardcode the spline's basis function columns
mtcars[c("ns_c1", "ns_c2", "ns_c3")] <- ns(mtcars$wt, df = 3)
mtcars[c("ns_d1", "ns_d2")] <- ns(mtcars$wt, df = 2)

x2 <- glmmTMB(
  am ~ ns_c1 + ns_c2 + ns_c3, 
  dispformula = ~ ns_d1 + ns_d2, 
  data = mtcars
)

# same coefficients
unlist(fixef(x)) == unlist(fixef(x2))
#>              cond.(Intercept)          cond.ns(wt, df = 3)1 
#>                          TRUE                          TRUE 
#>          cond.ns(wt, df = 3)2          cond.ns(wt, df = 3)3 
#>                          TRUE                          TRUE 
#>              disp.(Intercept) disp.splines::ns(wt, df = 2)1 
#>                          TRUE                          TRUE 
#> disp.splines::ns(wt, df = 2)2 
#>                          TRUE

# Convert the new x values into the spline's basis functions
newdata[c("ns_c1", "ns_c2", "ns_c3")] <- predict(ns(mtcars$wt, df = 3), newdata$wt)
newdata[c("ns_d1", "ns_d2")] <- predict(ns(mtcars$wt, df = 2), newdata$wt)

# same predictions
predict(x2, newdata) == predict(x, newdata)
#>  [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
#> [16] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
#> [31] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
#> [46] TRUE TRUE TRUE TRUE TRUE
```

**Comment by bbolker** [comment]
The workaround may be a useful clue about how to get this to work without complete refactoring ...

**Comment by bbolker** [comment]
Can you please `remotes::install_github("glmmTMB/glmmTMB/glmmTMB@more_predvars")` and check if this resolves the issue?  Further crazy test cases welcome ...

**Comment by bbolker** [comment]
@tjmahr Can you comment on whether this is fixed in https://github.com/glmmTMB/glmmTMB/pull/854 ?

**Comment by bbolker** [comment]
Presumably closed in #854.
